File: Javascript/114. Flatten Binary Tree to Linked List.js
/*
Given the root of a binary tree, flatten the tree into a "linked list":

The "linked list" should use the same TreeNode class where the right child pointer points to the next node in the list 
    and the left child pointer is always null.
The "linked list" should be in the same order as a pre-order traversal of the binary tree.
*/
function TreeNode(val, left, right) {
    this.val = (val===undefined ? 0 : val)
    this.left = (left===undefined ? null : left)
    this.right = (right===undefined ? null : right)
}
function printNode(node) {
    if (node === null) { return; }
    console.log(node.val);
    //printNode(node.left);
    printNode(node.right);
}  
/**
 * 114. Flatten Binary Tree to Linked List
 * @param {TreeNode} root
 * @return {void} Do not return anything, modify root in-place instead.
 */
var flatten = function(root) {
    let current = root;
    while (current !== null) {
        if (current.left !== null) {
            let predecessor = current.left;
            while (predecessor.right !== null) {
                predecessor = predecessor.right;
            }
            predecessor.right = current.right;
            // Move left to the right
            current.right = current.left;
            // Set all lefts to null
            current.left = null;
        }
        current = current.right;
    }
};
const root = new TreeNode(1,
    new TreeNode(2, new TreeNode(3), new TreeNode(4)),
    new TreeNode(5, null, new TreeNode(6))
);

flatten(root); 
printNode(root); // Expect [1,null,2,null,3,null,4,null,5,null,6]

File: Rust/114_Flatten_Binary_Tree_to_Linked_List.rs
/*
Given the root of a binary tree, flatten the tree into a "linked list":

The "linked list" should use the same TreeNode class where the right child pointer points to the next node in the list 
    and the left child pointer is always null.
The "linked list" should be in the same order as a pre-order traversal of the binary tree.
    Uses in-place Morris traversal
*/
/* 114. Flatten Binary Tree to Linked List */
use std::rc::Rc;
use std::cell::RefCell;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32, left: Option<Rc<RefCell<TreeNode>>>, right: Option<Rc<RefCell<TreeNode>>>) -> Rc<RefCell<TreeNode>> {
        Rc::new(RefCell::new(TreeNode { val, left, right }))
    }
}

struct Solution;
impl Solution {
    pub fn flatten(root: &mut Option<Rc<RefCell<TreeNode>>>) {
        let mut current = root.clone();
        
        while let Some(node_rc) = current {
            let mut node = node_rc.borrow_mut();
            
            if let Some(left_rc) = node.left.take() {
                let mut predecessor = left_rc.clone();
                
                while predecessor.borrow().right.is_some() {
                    let right_child = predecessor.borrow().right.clone();
                    predecessor = right_child.unwrap();;
                }
                predecessor.borrow_mut().right = node.right.take();
                node.right = Some(left_rc);
            }    
            current = node.right.clone();
        }
    }
}
fn main() {
    let root = TreeNode::new(
        1, 
        Some(TreeNode::new(
            2, 
            Some(TreeNode::new(3, None, None)), 
            Some(TreeNode::new(4, None, None)),
        )),
        Some(TreeNode::new(
            5, 
            None, 
            Some(TreeNode::new(6, None, None)),
        )),
    );
    let mut root_option = Some(root.clone());
    Solution::flatten(&mut root_option);
    println!("{:?}", root_option); // Expect [1,null,2,null,3,null,4,null,5,null,6]
}